### PR TITLE
[P2LB] Fix for banner text to not `strip_tags` for all markup

### LINKED ISSUE #7018: P2LB banners converted with buttons will not render basic markup tags
# Problem
When P2LB converts a section into a banner and there is a link, the paragraph won't render `<p>`,`<strong>`,`<em>`, `<h*>` tags in the banner. 

# Steps to reproduce
```
ddev blt ds --site brand.uiowa.edu && ddev drush @brand.local config-split:activate p2lb -y && ddev drush @brand.local cim -y && ddev drush @brand.local uli admin/content/sitenow-converter
```
```
ddev blt ds --site mnh.uiowa.edu && ddev drush @mnh.local config-split:activate p2lb -y && ddev drush @mnh.local cim -y && ddev drush @mnh.local uli admin/content/sitenow-converter
```
1. Convert the home page
2. Watch how the paragraph text will not render with `<p>` tags. 

Screenshot 2023-10-18 at 11 50 18 AM

# Proposed solution
Adjust https://github.com/uiowa/uiowa/blob/main/docroot/modules/custom/sitenow_p2lb/sitenow_p2lb.module#L1043-L1071.

### PR DESCRIPTION
**Pull request by bspeare**
Resolves https://github.com/uiowa/uiowa/issues/7018. 

```
ddev blt ds --site brand.uiowa.edu && ddev drush @brand.local config-split:activate p2lb -y && ddev drush @brand.local cim -y && ddev drush @brand.local uli admin/content/sitenow-converter
```
```
ddev blt ds --site mnh.uiowa.edu && ddev drush @mnh.local config-split:activate p2lb -y && ddev drush @mnh.local cim -y && ddev drush @mnh.local uli admin/content/sitenow-converter
```
```
ddev blt ds --site hawkeyemarchingband.uiowa.edu && ddev drush @hawkeyemarchingband.local config-split:activate p2lb -y && ddev drush @hawkeyemarchingband.local cim -y && ddev drush @hawkeyemarchingband.local uli admin/content/sitenow-converter
```
```
ddev blt ds --site paleorepository.sites.uiowa.edu && ddev drush @sitespaleorepository.local config-split:activate p2lb -y && ddev drush @sitespaleorepository.local cim -y && ddev drush @sitespaleorepository.local uli admin/content/sitenow-converter
```

```
ddev blt ds --site latinxcouncil.uiowa.edu && ddev drush @latinxcouncil.local config-split:activate p2lb -y && ddev drush @latinxcouncil.local cim -y && ddev drush @latinxcouncil.local uli admin/content/sitenow-converter
```

1. Convert the home page and checkout out all the banners. 
2. Watch how the paragraph text will now render with `<p>` tags. 
3. Bonus: use kitchen sink HTML from details below

<details>

```
<h3>Typography styles</h3>

<p>Here are some headline and typography&nbsp;options that you can use in copy. Click the "Configure" link at the top right of this block, copy&nbsp;your preferred style from the editor, paste to the page you're working on, and edit the text to suit your message.&nbsp;</p>

<h2 class="bold-headline bold-headline--caps"><span class="heading__heading">Bold Headline</span></h2>

<hr class="element--spacer-thin" />
<h2 class="bold-headline bold-headline--serif"><span class="heading__heading">Bold headline serif</span></h2>

<hr class="element--spacer-thin" />
<h2 class="bold-headline--underline bold-headline bold-headline--serif">Bold headline serif underline</h2>

<hr class="element--spacer-thin" />
<p class="uids-component--light-intro">Light intro type for body and paragraph text</p>

<hr class=".element--spacer-separator " />
<p class="uids-component--bold-intro">Bold intro type for body and paragraph text.</p>

<p class="uids-component--bold-intro">Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nulla accumsan, metus ultrices eleifend gravida, nulla nunc varius lectus, nec rutrum justo nibh eu lectus. Ut vulputate semper dui. Fusce erat odio, sollicitudin vel erat vel, interdum mattis neque.</p>

<p class="uids-component--light-intro">Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nulla accumsan, metus ultrices eleifend gravida, nulla nunc varius lectus, nec rutrum justo nibh eu lectus. Ut vulputate semper dui. Fusce erat odio, sollicitudin vel erat vel, interdum mattis neque.</p>

<p>This paragraph is nested inside an article. It contains many different, sometimes useful, <a href="https://www.w3schools.com/tags/">HTML5 tags</a>. Of course there are classics like <em>emphasis</em>, <strong>strong</strong>, and small but there are many others as well. Hover the following text for abbreviation tag: abbr. Similarly, you can use acronym tag like this: ftw. You can define deleted text which often gets replaced with inserted text.</p>

<p>You can also use keyboard text, which sometimes is styled similarly to the <code>&lt;code&gt;</code> or samp tags. Even more specifically, there is a tag just for variables. Not to be mistaken with blockquotes below, the quote tag lets you denote something as quoted text. Lastly don't forget the sub (H<sub>2</sub>O) and sup (E = MC<sup>2</sup>) tags.</p>

<p>Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nulla accumsan, metus ultrices eleifend gravida, nulla nunc varius lectus, nec rutrum justo nibh eu lectus. Ut vulputate semper dui. Fusce erat odio, sollicitudin vel erat vel, interdum mattis neque.</p>

<p>At vero eos et accusamus et iusto odio dignissimos ducimus qui blanditiis praesentium voluptatum deleniti atque corrupti quos dolores et quas molestias excepturi sint occaecati cupiditate non provident, similique sunt in culpa qui officia deserunt mollitia animi, id est laborum et dolorum fuga. Et harum quidem rerum facilis est et expedita distinctio. Nam libero tempore, cum soluta nobis est eligendi optio cumque nihil impedit quo minus id quod maxime placeat facere possimus, omnis voluptas assumenda est, omnis dolor repellendus. Temporibus autem quibusdam et aut officiis debitis aut rerum necessitatibus saepe eveniet ut et voluptates repudiandae sint et molestiae non recusandae. Itaque earum rerum hic tenetur a sapiente delectus, ut aut reiciendis voluptatibus maiores alias consequatur aut perferendis doloribus asperiores repellat.</p>

<h2>Heading 2</h2>

<h3>Heading 3</h3>

<h4>Heading 4</h4>

<h5>Heading 5</h5>

<h6>Heading 6</h6>

<p>Curabitur accumsan turpis pharetra <strong>augue tincidunt</strong> blandit. Quisque condimentum maximus mi, sit amet commodo arcu rutrum id. Proin pretium urna vel cursus venenatis. Suspendisse potenti. Etiam mattis sem rhoncus lacus dapibus facilisis. Donec at dignissim dui. Ut et neque nisl.</p>

<p>Pellentesque habitant morbi tristique senectus et netus et malesuada fames ac turpis egestas. Vestibulum tortor quam, feugiat vitae, ultricies eget, tempor sit amet, ante. Donec eu libero sit amet quam egestas semper. Aenean ultricies mi vitae est. Mauris placerat eleifend leo.</p>

<p><a class="bttn bttn--tertiary" href="https://www.example.com">More News <span class="fa-arrow-right fas"></span> </a> <a class="bttn bttn--primary" href="https://www.example.com"> More News <span class="fa-arrow-right fas"></span> </a> <a class="bttn bttn--secondary" href="https://www.example.com"> More News <span class="fa-arrow-right fas"></span> </a> <!-- Start: @button--button-outline --> <a class="bttn bttn--transparent" href="https://www.example.com"> More News <span class="fa-arrow-right fas"></span> </a> <!-- Start: @button--button-outline-sans-serif --> <a class="bttn bttn--transparent bttn--light-font" href="https://www.example.com"> More News <span class="fa-arrow-right fas"></span> </a> <a class="bttn bttn--secondary bttn--light-font" href="https://www.example.com"> More News </a> <a class="bttn bttn--transparent borderless bttn--light-font" href="https://www.example.com"> More News <span class="fa-arrow-right fas"></span> </a></p>

<p>&nbsp;</p>

<div class="alert alert--success">Danger Alert with <a href="https://example.com">link</a></div>

<div class="alert alert--danger">Danger Alert with <a href="https://example.com">link</a></div>

<div class="alert alert--warning">Danger Alert with <a href="https://example.com">link</a></div>

<div class="alert alert--info">Danger Alert with <a href="https://example.com">link</a></div>

<p>Praesent egestas tristique nibh. Quisque malesuada placerat nisl. Etiam ut purus mattis mauris sodales aliquam. Curabitur vestibulum aliquam leo..</p>

<blockquote>
<p>Nam at tortor in tellus interdum sagittis. Morbi vestibulum volutpat enim.</p>
</blockquote>

<p>&nbsp;</p>

<p>Donec vitae sapien ut libero venenatis faucibus. Nam quam nunc, blandit vel, luctus pulvinar, hendrerit id, lorem. Nullam nulla eros, ultricies sit amet, nonummy id, imperdiet feugiat, pede.</p>

<table>
	<caption>Tables can have captions now.</caption>
	<thead>
		<tr>
			<th>Person</th>
			<th>Number</th>
			<th>Third Column</th>
		</tr>
	</thead>
	<tbody>
		<tr>
			<th>Someone Lastname</th>
			<td>900</td>
			<td>Nullam quis risus eget urna mollis ornare vel eu leo.</td>
		</tr>
		<tr>
			<th><a href="#">Person Name</a></th>
			<td>1200</td>
			<td>Vestibulum id ligula porta felis euismod semper. Donec ullamcorper nulla non metus auctor fringilla.</td>
		</tr>
		<tr>
			<th>Another Person</th>
			<td>1500</td>
			<td>Vivamus sagittis lacus vel augue laoreet rutrum faucibus dolor auctor. Nullam id dolor id nibh ultricies vehicula ut id elit.</td>
		</tr>
		<tr>
			<th>Last One</th>
			<td>2800</td>
			<td>Morbi leo risus, porta ac consectetur ac, vestibulum at eros. Cras mattis consectetur purus sit amet fermentum.</td>
		</tr>
	</tbody>
</table>

<p>&nbsp;</p>

<p>&nbsp;</p>

<table>
	<caption>Tables can have captions now.</caption>
	<thead>
		<tr>
			<th>Person</th>
			<th>Number</th>
			<th>Third Column</th>
			<th>Fourth Column</th>
			<th>Fifth Column</th>
			<th>Sixth Column</th>
			<th>Seventhth Column</th>
			<th>Eighth Column</th>
			<th>Ninth Column</th>
			<th>Tenth Column</th>
			<th>Eleventh Column</th>
		</tr>
	</thead>
	<tbody>
		<tr>
			<th>Someone Lastname</th>
			<td>900</td>
			<td>9999</td>
			<td>1999</td>
			<td>1999</td>
			<td>1999</td>
			<td>1999</td>
			<td>1999</td>
			<td>1999</td>
			<td>1999</td>
			<td>1999</td>
		</tr>
		<tr>
			<th><a href="#">Person Name</a></th>
			<td>1200</td>
			<td>9999</td>
			<td>1999</td>
			<td>1999</td>
			<td>1999</td>
			<td>1999</td>
			<td>1999</td>
			<td>1999</td>
			<td>1999</td>
			<td>1999</td>
		</tr>
		<tr>
			<th>Another Person</th>
			<td>1500</td>
			<td>9999</td>
			<td>1999</td>
			<td>1999</td>
			<td>1999</td>
			<td>1999</td>
			<td>1999</td>
			<td>1999</td>
			<td>1999</td>
			<td>1999</td>
		</tr>
		<tr>
			<th>some other person</th>
			<td>2800</td>
			<td>9999</td>
			<td>1999</td>
			<td>1999</td>
			<td>1999</td>
			<td>1999</td>
			<td>1999</td>
			<td>1999</td>
			<td>1999</td>
			<td>1999</td>
		</tr>
		<tr>
			<th>Someone and a Lastname</th>
			<td>500</td>
			<td>9999</td>
			<td>1999</td>
			<td>1999</td>
			<td>1999</td>
			<td>1999</td>
			<td>1999</td>
			<td>1999</td>
			<td>1999</td>
			<td>1999</td>
		</tr>
		<tr>
			<th><a href="#">Linked Name</a></th>
			<td>1720</td>
			<td>9999</td>
			<td>1999</td>
			<td>1999</td>
			<td>1999</td>
			<td>1999</td>
			<td>1999</td>
			<td>1999</td>
			<td>1999</td>
			<td>1999</td>
		</tr>
		<tr>
			<th>Another random person</th>
			<td>1200</td>
			<td>9999</td>
			<td>1999</td>
			<td>1999</td>
			<td>1999</td>
			<td>1999</td>
			<td>1999</td>
			<td>1999</td>
			<td>1999</td>
			<td>1999</td>
		</tr>
		<tr>
			<th>Some other Person</th>
			<td>2200</td>
			<td>9999</td>
			<td>1999</td>
			<td>1999</td>
			<td>1999</td>
			<td>1999</td>
			<td>1999</td>
			<td>1999</td>
			<td>1999</td>
			<td>1999</td>
		</tr>
		<tr>
			<th>Firstname and something</th>
			<td>920</td>
			<td>9999</td>
			<td>1999</td>
			<td>1999</td>
			<td>1999</td>
			<td>1999</td>
			<td>1999</td>
			<td>1999</td>
			<td>1999</td>
			<td>1999</td>
		</tr>
		<tr>
			<th><a href="#">Linked Other Name</a></th>
			<td>1208</td>
			<td>9999</td>
			<td>1999</td>
			<td>1999</td>
			<td>1999</td>
			<td>1999</td>
			<td>1999</td>
			<td>1999</td>
			<td>1999</td>
			<td>1999</td>
		</tr>
		<tr>
			<th>Another Person Randomly selected</th>
			<td>1550</td>
			<td>9999</td>
			<td>1999</td>
			<td>1999</td>
			<td>1999</td>
			<td>1999</td>
			<td>1999</td>
			<td>1999</td>
			<td>1999</td>
			<td>1999</td>
		</tr>
		<tr>
			<th>Last One</th>
			<td>2830</td>
			<td>9999</td>
			<td>1999</td>
			<td>1999</td>
			<td>1999</td>
			<td>1999</td>
			<td>1999</td>
			<td>1999</td>
			<td>1999</td>
			<td>1999</td>
		</tr>
	</tbody>
</table>

<p>&nbsp;</p>

<p>&nbsp;</p>

<p>Woodlands payment Osgiliath tightening. Barad-dur follow belly comforts tender tough bell? Many that live deserve death. Some that die deserve life. Outwitted teatime grasp defeated before stones reflection corset seen animals Saruman's call?</p>

<h2>Tad survive ensnare joy mistake courtesy Bagshot Row.</h2>

<p>Ligulas step drops both? You shall not pass! Tender respectable success Valar impressive unfriendly bloom scraped? Branch hey-diddle-diddle pony trouble'll sleeping during jump Narsil.</p>

<h3>North valor overflowing sort Iáve mister kingly money?</h3>

<p>Curse you and all the halflings! Deserted anytime Lake-town burned caves balls. Smoked lthilien forbids Thrain?</p>

<ul>
	<li>In fermentum leo eu lectus mollis, quis dictum mi aliquet.</li>
	<li>Morbi eu nulla lobortis, lobortis est in, fringilla felis.</li>
	<li>Aliquam nec felis in sapien venenatis viverra fermentum nec lectus.</li>
	<li>Ut non enim metus.
	<ul>
		<li>In fermentum leo eu lectus mollis, quis dictum mi aliquet.</li>
		<li>Morbi eu nulla lobortis, lobortis est in, fringilla felis.</li>
		<li>Aliquam nec felis in sapien venenatis viverra fermentum nec lectus.</li>
		<li>Ut non enim metus.
		<ul>
			<li>In fermentum leo eu lectus mollis, quis dictum mi aliquet.</li>
			<li>Morbi eu nulla lobortis, lobortis est in, fringilla felis.</li>
			<li>Aliquam nec felis in sapien venenatis viverra fermentum nec lectus.</li>
			<li>Ut non enim metus.</li>
		</ul>
		</li>
	</ul>
	</li>
</ul>

<ol>
	<li>In fermentum leo eu lectus mollis, quis dictum mi aliquet.</li>
	<li>Morbi eu nulla lobortis, lobortis est in, fringilla felis.</li>
	<li>Aliquam nec felis in sapien venenatis viverra fermentum nec lectus.</li>
	<li>Ut non enim metus.
	<ol>
		<li>In fermentum leo eu lectus mollis, quis dictum mi aliquet.</li>
		<li>Morbi eu nulla lobortis, lobortis est in, fringilla felis.</li>
		<li>Aliquam nec felis in sapien venenatis viverra fermentum nec lectus.</li>
		<li>Ut non enim metus.
		<ol>
			<li>In fermentum leo eu lectus mollis, quis dictum mi aliquet.</li>
			<li>Morbi eu nulla lobortis, lobortis est in, fringilla felis.</li>
			<li>Aliquam nec felis in sapien venenatis viverra fermentum nec lectus.</li>
			<li>Ut non enim metus.</li>
		</ol>
		</li>
	</ol>
	</li>
</ol>

<ul class="element--circle-list ">
	<li>In fermentum leo eu lectus mollis, quis dictum mi aliquet.</li>
	<li>Morbi eu nulla lobortis, lobortis est in, fringilla felis.</li>
	<li>Aliquam nec felis in sapien venenatis viverra fermentum nec lectus.</li>
	<li>Ut non enim metus.</li>
</ul>

<ol>
	<li>In fermentum leo eu lectus mollis, quis dictum mi aliquet.</li>
	<li>Morbi eu nulla lobortis, lobortis est in, fringilla felis.</li>
	<li>Aliquam nec felis in sapien venenatis viverra fermentum nec lectus.</li>
	<li>Ut non enim metus.</li>
</ol>

<ol class="element--circle-list ">
	<li>Complete <a href="https://admissions.uiowa.edu/academics-undergraduate-programs-undergraduate/minimum-high-school-course-requirements">high school course requirements</a>.&nbsp;&nbsp;</li>
	<li>Achieve the minimum required score on the <a href="https://admissions.uiowa.edu/rai">Regents Admissions Index (RAI)</a>.&nbsp;</li>
	<li>Meet <a href="https://admissions.uiowa.edu/academics/first-year-admission">admissions requirements</a> for your college of choice.&nbsp;&nbsp;</li>
	<li>Submit your <a href="https://financialaid.uiowa.edu/apply">FAFSA</a> by October 1st and explore financial aid.&nbsp;&nbsp;</li>
	<li>Complete an online application.&nbsp;&nbsp;</li>
	<li>
	<p>This paragraph is nested inside an article. It contains many different, sometimes useful, <a href="https://www.w3schools.com/tags/">HTML5 tags</a>. Of course there are classics like <em>emphasis</em>, <strong>strong</strong>, and small but there are many others as well. Hover the following text for abbreviation tag: abbr. Similarly, you can use acronym tag like this: ftw. You can define deleted text which often gets replaced with inserted text.</p>

	<p>You can also use keyboard text, which sometimes is styled similarly to the <code>&lt;code&gt;</code> or samp tags. Even more specifically, there is a tag just for variables. Not to be mistaken with blockquotes below, the quote tag lets you denote something as quoted text. Lastly don't forget the sub (H<sub>2</sub>O) and sup (E = MC<sup>2</sup>) tags.</p>
	</li>
</ol>

<dl>
	<dt>Definition List Title</dt>
	<dd>Definition list division.</dd>
	<dt>Kitchen Sink</dt>
	<dd>Used in expressions to describe work in which all conceivable (and some inconceivable) sources have been mined. In this case, a bunch of markup.</dd>
	<dt>aside</dt>
	<dd>Defines content aside from the page content</dd>
	<dt>blockquote</dt>
	<dd>Defines a section that is quoted from another source</dd>
</dl>

```
</details>

### CODE REVIEW

**Comment by bspeare** [comment]
Maybe we don't want to `strip_tags` on the banner text conversion?   

Before:
Screenshot 2023-10-18 at 1 56 55 PM

After:

Screenshot 2023-10-18 at 1 53 54 PM

**Comment by joewhitsitt** [comment]
Do you recall why we wanted to strip_tags in the beginning? Are there things in filtered_html that we don't want brought over to the banner?

Wonder if @cory-skeers can weigh in with background from https://github.com/uiowa/uiowa/pull/5770 where it was introduced.

We could instead expand strip_tags to allow certain tags. https://www.php.net/manual/en/function.strip-tags.php

**Comment by cory-skeers** [comment]
> Do you recall why we wanted to strip_tags in the beginning? Are there things in filtered_html that we don't want brought over to the banner?
> 
> Wonder if @cory-skeers can weigh in with background from #5770 where it was introduced.
>
> We could instead expand strip_tags to allow certain tags. https://www.php.net/manual/en/function.strip-tags.php

https://github.com/uiowa/uiowa/pull/5770#issuecomment-1270550727

There were some encoding issues, and the full `strip_tags` was probably a heavy handed approach at it that could be lightened up to allow tags that would be allowed in minimal.

**Comment by bspeare** [comment]
The text format does strip out unwanted tags like `<table>` and others when it is converted to the banners text format.

File: docroot/modules/custom/sitenow_p2lb/sitenow_p2lb.module [hunk 1]
```diff
@@ -1057,8 +1057,7 @@ function sitenow_p2lb_section_image($section_image_fid, $banner_text, $node) {
   // Assumes only one button or link is present.
   // Any more links will remain in the body.
   if (preg_match('|(.*?)(<a.*?href="(.*?)".*?>(.*?)<\/a>)(.*)|is', $text['value'], $matches)) {
-    // Attach the bookended content into a single value.
-    $text['value'] = strip_tags($matches[1] . $matches[5]);
+    $text['value'] = $matches[1] . $matches[5];
```
Comment (joewhitsitt) [inline docroot/modules/custom/sitenow_p2lb/sitenow_p2lb.module]:
```suggestion
    $text['value'] = strip_tags($matches[1] . $matches[5], ['p', 'a', 'strong', 'em', 'h2', 'h3', 'h4','h6' ]);
```

>as of PHP 7.4.0 the line above can be written as: echo strip_tags($text, ['p', 'a']);

not tested. [updated with Ben's suggestion](https://github.com/uiowa/uiowa/pull/7019/files#r1370663342)

**Comment by cory-skeers** [comment]
> The text format does strip out unwanted tags like `<table>` and others when it is converted to the banners text format.

It might've been something like we've run into in the past with migrations, where the problem was the text format not being properly set. In the case of migrations, once the page was saved through a node edit, the bad format would get fixed. If this was something similar, maybe we've fixed the underlying issue elsewhere.

In either case, I think removing the `strip_tags` is probably fine.

**Comment by joewhitsitt** [comment]
i'm indifferent. i am just cautious here and `p` tags is what the bug issue is about.

File: docroot/modules/custom/sitenow_p2lb/sitenow_p2lb.module (hunk 1 above)
Comment (bspeare) [reply docroot/modules/custom/sitenow_p2lb/sitenow_p2lb.module]: Thanks @joewhitsitt, I think we want something more like `$text['value'] = strip_tags($matches[1] . $matches[5], ['p', 'a', 'strong', 'em', 'h2', 'h3', 'h4','h6' ]);` that we can update as we go.  How does that sound?  I can post example screenshots if you'd like to see some samples.

**Comment by bspeare** [comment]
> i'm indifferent. i am just cautious here and `p` tags is what the bug issue is about.

I appreciate your replies! I updated the the bug report to include more of the markup that is being stripped out.

File: docroot/modules/custom/sitenow_p2lb/sitenow_p2lb.module (hunk 1 above)
Comment (joewhitsitt) [reply docroot/modules/custom/sitenow_p2lb/sitenow_p2lb.module]: updated the commit suggestion based on your comment. again, i haven't tested anything yet.

**Comment by joewhitsitt** [comment]
pulling to test

**Review by joewhitsitt — APPROVED** [review]
This works for me. Tested on a few.

**Comment by bspeare** [comment]
Thanks @joewhitsitt!

**Pull request merged**